src/UserSpaceInstrumentation/AccessTraceesMemory.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AccessTraceesMemory.h"
+
+#include <absl/base/casts.h>
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_format.h>
+#include <absl/strings/str_split.h>
+#include <sys/ptrace.h>
+
+#include <cerrno>
+#include <cstring>
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/SafeStrerror.h"
+
+namespace orbit_user_space_instrumentation {
+
+using orbit_base::ReadFileToString;
+
+[[nodiscard]] ErrorMessageOr<std::vector<uint8_t>> ReadTraceesMemory(pid_t pid,
+                                                                     uint64_t address_start,
+                                                                     uint64_t length) {
+  // Round up length to next multiple of eight.
+  length = ((length + 7) / 8) * 8;
+  std::vector<uint8_t> bytes(length);
+  for (size_t i = 0; i < length / 8; i++) {
+    const uint64_t data = ptrace(PTRACE_PEEKDATA, pid, address_start + i * 8, NULL);
+    if (errno) {
+      return ErrorMessage(
+          absl::StrFormat("Failed to PTRACE_PEEKDATA for pid: %d with errno %d: \"%s\"", pid, errno,
+                          SafeStrerror(errno)));
+    }
+    std::memcpy(bytes.data() + (i * sizeof(uint64_t)), &data, 8);
+  }
+  return bytes;
+}
+
+[[nodiscard]] ErrorMessageOr<void> WriteTraceesMemory(pid_t pid, uint64_t address_start,
+                                                      const std::vector<uint8_t>& bytes) {
+  size_t pos = 0;
+  do {
+    // Pack 8 bytes for writing into `data`.
+    uint64_t data = 0;
+    std::memcpy(&data, bytes.data() + pos, sizeof(uint64_t));
+    if (ptrace(PTRACE_POKEDATA, pid, address_start + pos, data) == -1) {
+      ErrorMessage(absl::StrFormat("Failed to PTRACE_POKEDATA for pid: %d.", pid));
```

Comment:
Isn't `errno` set even if `ptrace` itself also returns a status code with `POKEDATA`?

Reply:
To my understanding errno is only set by the  PTRACE_PEEK* stuff since the retunr value is user for the actual data. At least this is how I understand this:

  https://man7.org/linux/man-pages/man2/ptrace.2.html

   _C library/kernel differences
       At the system call level, the PTRACE_PEEKTEXT, PTRACE_PEEKDATA,
       and PTRACE_PEEKUSER requests have a different API: they store the
       result at the address specified by the data parameter, and the
       return value is the error flag.  The glibc wrapper function
       provides the API given in DESCRIPTION above, with the result
       being returned via the function return value._

Reply:
Yep, sounds like it.